Admin Panel for Popularity Transfers

advay26 opened the pull request:
## Admin Panel for Popularity Transfers

Addresses https://github.com/NuGet/Engineering/issues/4746

![image](https://github.com/NuGet/NuGetGallery/assets/82980589/3a6ef4ec-da95-483c-bad5-89c51673afd1)

<details><summary>You can now enter a list of packages you want to transfer popularity from, and a list of packages you want to transfer to. </summary>

![image](https://github.com/NuGet/NuGetGallery/assets/82980589/4c162811-b05c-4460-9517-9351f5f52ae8)
</details> 

<details><summary>It then validates your input. </summary>

![image](https://github.com/NuGet/NuGetGallery/assets/82980589/265a5abb-3199-4e48-9005-ea26e7e660df)
</details> 

<details><summary>Submitting your changes updates the DB with new entries in the PackageRenames table.</summary>

![image](https://github.com/NuGet/NuGetGallery/assets/82980589/adcaf6c3-35ff-4788-95a3-75e84a443df2)

You can see the new Serilog (Key = 11) --> Moq (Key = 9) entry
![image](https://github.com/NuGet/NuGetGallery/assets/82980589/6a315e46-0686-4776-be62-56fefa752dee)
</details> 

<details><summary>Error Handling</summary>

![image](https://github.com/NuGet/NuGetGallery/assets/82980589/5d5dc90a-b10f-4884-80f0-a51795ef7a83)

![image](https://github.com/NuGet/NuGetGallery/assets/82980589/dea204c9-20bc-44c1-8691-7438c1c4418a)
</details> 

<details><summary>Checking for existing popularity transfers</summary>

**NOTE:** Our current playbook asks us to delete any existing popularity transfers for the 'From' packages, so the popularity transfer operation deletes previous entries from the DB. Let me know if this doesn't sound right, or we want to give the DRIs more information/options on what to do with existing popularity transfers. Currently, we just show a warning when we validate the input, and then if a user submits the changes, then we remove all previous entries.
![image](https://github.com/NuGet/NuGetGallery/assets/82980589/c9130282-d647-40e1-93fd-20139771e7c7)

We also show a warning if one of the 'To' packages has existing renames, as this would result in a transitive popularity transfer relationship (a -> b -> c). The DRI can then look at the table in the DB to see if they still want to proceed.
![image](https://github.com/NuGet/NuGetGallery/assets/82980589/a126aa44-0d18-46dc-bfea-70de63af78fb)

</details> 

<details><summary>Multiple packages at once</summary>

![image](https://github.com/NuGet/NuGetGallery/assets/82980589/fd73204e-b628-4dc3-a0c0-59a8cfc0f51b)

![image](https://github.com/NuGet/NuGetGallery/assets/82980589/48ceb4d5-6c8a-4965-b010-aa0b95111466)
</details>

advay26 commented on src/NuGetGallery/Areas/Admin/Views/ReservedNamespace/Index.cshtml:
```diff
@@ -1,6 +1,6 @@
 ﻿@model ReservedNamespaceViewModel
 @{
-    ViewBag.Title = "Index";
+    ViewBag.Title = "Reserved Namespaces";
```
This is separate to the Popularity Transfers changes.

I noticed that the Reserved Namespaces page just says "Index" in the title and wanted to make it more specific.

drewgillies commented on src/NuGetGallery/Areas/Admin/Controllers/PopularityTransferController.cs:
```diff
@@ -0,0 +1,175 @@
+
+                if (packageFrom is null)
+                {
+                    var InvalidId = packagesFromInput
```
Couldn't the population of the list be a one-time process before the for loop, and then the resultant array just be indexed here?

advay26 replied:
You're right. I've changed it now so that the list only has to be populated once

drewgillies commented on src/NuGetGallery/Areas/Admin/Controllers/PopularityTransferController.cs:
```diff
@@ -0,0 +1,175 @@
+                }
+                if (packageTo is null)
+                {
+                    var InvalidId = packagesToInput
```
Couldn't the population of the list be a one-time process before the for loop, and then the resultant array just be indexed here?

drewgillies reviewed (dismissed):
Looks great, Advay--my comments are really just perf nits.

erdembayar commented:
So far I only see 1 to 1 transfer screenshot.
Could you please include 1 screenshot for 2 packages to 2 packages preview window?

Do we do any validation such as if there's any circular transfer mapping?
Let's say there're `A`, `B`, `C` packages and there's already `A->B and B-> C` mapping in package rename table.
What happens if mistakenly add `C->A` mapping now? Does it warn if scenario happen?

advay26 commented:
> So far I only see 1 to 1 transfer screenshot. Could you please include 1 screenshot for 2 packages to 2 packages preview window?

@erdembayar I've added new screenshots showing the transfer of multiple packages https://github.com/NuGet/NuGetGallery/pull/9668#issue-1903788734

> Do we do any validation such as if there's any circular transfer mapping? Let's say there're `A`, `B`, `C` packages and there's already `A->B and B-> C` mapping in package rename table. What happens if mistakenly add `C->A` mapping now? Does it warn if scenario happen?

We don't currently have any checks for this in the OneNote, but I imagine we should. I added a warning if one of the 'To' packages has existing popularity transfers. If the DRI sees this, they can look at the DB and see the related rows to decide how they want to proceed. I added a screenshot for this too.

erdembayar commented on src/NuGetGallery/Areas/Admin/Controllers/PopularityTransferController.cs:
```diff
@@ -0,0 +1,196 @@
+                var previousRenames = _packageRenameService.GetPackageRenames(_packageService.FindPackageRegistrationById(input.FromId));
+                previousPackageRenames.AddRange(previousRenames);
+
+                if (input.FromId.Length > maxIdLength)
```
Probably `maxIdLength = Math.Max(maxIdLength, input.FromId.Length;);` is more concise and common syntax for this case.

erdembayar commented on src/NuGetGallery/Areas/Admin/Controllers/PopularityTransferController.cs:
```diff
@@ -0,0 +1,196 @@
+
+            _packageRenameRepository.DeleteOnCommit(previousPackageRenames);
+            _packageRenameRepository.InsertOnCommit(newPackageRenames);
+            await _packageRenameRepository.CommitChangesAsync();
```
Do we use `transaction` or `unit of work` pattern in our code? 
If case some changes went through but another steps failed later then we can revert whole transaction.

advay26 replied:
We discussed this on the Server channel too where @joelverhagen explained how we use Entity Framework, but the summary was that we do use both `unit of work` and `transaction` patterns. All changes to EF's DB context (`_entitiesContext` in the code here) are bunched together (`InsertOnCommit`, `DeleteOnCommit`, etc.) and then applied at once using `SaveChangesAsync`. The 'SaveChangesAsync` abstracts away the transaction, but it does use it internally when applying the SQL changes to the DB. If any of the commands in the change-set fail, it rolls back the transaction instead of committing.

erdembayar commented on src/NuGetGallery/Areas/Admin/Views/PopularityTransfer/Index.cshtml:
```diff
@@ -0,0 +1,167 @@
+    </div>
+</section>
+
+@section BottomScripts {
```
Does `BottomScripts` mean it's executed after page rendered? How is it wired?

advay26 replied:
Yes, BottomScripts get loaded into the html in a `<script>` tag. It's placed within a `document.ready()` function and is run when the DOM is loaded.

erdembayar reviewed (approved):
Few comments, not blocking.

joelverhagen commented on src/NuGetGallery/Areas/Admin/Controllers/PopularityTransferController.cs:
```diff
@@ -0,0 +1,196 @@
+                }
+
+                // check for duplicate package ids
+                if (inputKeys.Contains(packageFrom.Key))
```
nit: the `Contains` and `Add` can be combined because `Add` returns `true` if it didn't exist before the `Add`. 
https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1.add?view=netframework-4.7.2#returns

joelverhagen commented on src/NuGetGallery/Areas/Admin/Controllers/PopularityTransferController.cs:
```diff
@@ -0,0 +1,196 @@
+                inputKeys.Add(packageTo.Key);
+
+                // create validated input result
+                var input = new PopularityTransferItem(CreatePackageSearchResult(packageFrom.Packages.First()),
```
nit: the `Packages` could be empty and `First` will fail. There exist package registrations with no packages (all deleted for example). Consider validating that `Packages.Count > 0` as well.

erdembayar replied:
@advay26 I can see unaddressed comment. After you address them then I can re-approve.

advay26 replied:
@erdembayar I added this check on line 73 and line 79 with `.Any()`, so I think I've addressed this too.

joelverhagen commented on src/NuGetGallery/Areas/Admin/Views/PopularityTransfer/Index.cshtml:
```diff
@@ -0,0 +1,167 @@
+            </table>
+
+            <div class="form-group col-md-12" style="display:none" data-bind="visible: existingPackageRenames().length > 0, foreach: existingPackageRenames">
+                @ViewHelpers.AlertInfo(@<text><span data-bind="text: $data"></span></text>)
```
nit: Warn not Info here

joelverhagen commented on src/NuGetGallery/Areas/Admin/Controllers/PopularityTransferController.cs:
```diff
@@ -0,0 +1,196 @@
+
+            foreach (var input in confirmedInputs)
+            {
+                message += $"\t{input.FromId.PadRight(maxIdLength)}  --->  {input.ToId}\n";
```
```suggestion
                message += $"\t{input.FromId.PadRight(maxIdLength)}  ➡️  {input.ToId}\n";
```

😅

joelverhagen replied:
It should work! :P

joelverhagen reviewed (dismissed):
Looks great. Nice work.

joelverhagen commented:
One thing to consider is to enhance the playbook or a minimal message in the UI to emphasize which validations are done by the tool and which are essentially left as optional in the playbook.

advay26 commented:
> One thing to consider is to enhance the playbook or a minimal message in the UI to emphasize which validations are done by the tool and which are essentially left as optional in the playbook.

That makes sense. I have a work item to update the OneNote page once this Admin Panel is live, so I'll keep that in mind and call it out there.
